### ISSUE TITLE
document the "insert" flag for event.listen and add example regarding setting flags, also mention this re: mysql sql_mode

### ISSUE THREAD

**Issue by sqlalchemy-bot** [sqlalchemy/sqlalchemy#2937]
**Migrated issue, originally created by Michael Bayer ([@zzzeek](https://github.com/zzzeek))**

    from sqlalchemy import create_engine, event
    
    eng = create_engine("mysql://scott:tiger@localhost/test", echo='debug')
    
    @event.listens_for(eng, "first_connect", insert=True)  # make sure we're the very first thing
    @event.listens_for(eng, "connect")
    def connect(dbapi_connection, connection_record):
        cursor = dbapi_connection.cursor()
        cursor.execute("SET sql_mode = 'STRICT_ALL_TABLES'")
    
    conn = eng.connect()

**Comment by jvanasco** [comment]
Current version: SQLAlchemy 1.4

The kwargs for `listens_for` are not documented.

In the various migration examples and tests we see:

* propagate=True
* retval=True
* insert=True
* once=True
* named=True

* the above intended example, is in metadata.rst, but for oracle:
   ```
   from sqlalchemy import event
   from sqlalchemy import create_engine

    engine = create_engine("oracle+cx_oracle://scott:tiger@tsn_name")

    @event.listens_for(engine, "connect", insert=True)
    def set_current_schema(dbapi_connection, connection_record):
        cursor = dbapi_connection.cursor()
        cursor.execute("ALTER SESSION SET CURRENT_SCHEMA=%s" % schema_name)
        cursor.close()
  ```   

I have no idea if the above list is complete, or what to document them as

**Comment by zzzeek** [comment]
named is here: https://docs.sqlalchemy.org/en/14/core/event.html?highlight=event%20listen#named-argument-styles

retval and propagate are event specific.

retval is introduced at https://docs.sqlalchemy.org/en/14/core/event.html?highlight=event%20listen#modifiers however this is generally documented where it is relevant, such as: https://docs.sqlalchemy.org/en/14/core/events.html?highlight=connection%20before_execute#sqlalchemy.events.ConnectionEvents.before_execute

propagate is relevant to ORM instruementation and mapping events: 

https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=after_configured#sqlalchemy.orm.MapperEvents.params.propagate

example: https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=after_configured#sqlalchemy.orm.MapperEvents.before_mapper_configured


"once" - i dont really want people to rely on this and id rather deprecate it

"insert" - this is easier to support however is still "dangerous" for end-user use, because SQLAlchemy itself also registers events in some cases and it would be better if end-users didnt rely upon ordering of event hooks.   this can be documented as a top level event kw with caveats mentioned.

**Comment by jvanasco** [comment]
Thanks! i can work with this.

**Comment by jvanasco** [comment]
@zzzeek It looks like `listens` shares the same kwargs as `listens_for` - as do all the events. Would it make sense to document `listens` and just reference the other event's kwargs to listens?

The way I'm building this out, is just doing a one-line description with the `:param bool name:` (copied from elsewhere in the docs) and then linking to elsewhere the docs for further reading.

**Comment by zzzeek** [comment]
is the question, "document only "listens", have "listens_for" just refer to "listens" for keyword docs" ?  that's fine

**Comment by jvanasco** [comment]
Yes. That's what I'll do.

**Comment by sqla-tester** [comment]
**jonathan vanasco** has proposed a fix for this issue in the **master** branch:

**Fixes: #2937** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3096
